src/appleseed/foundation/utility/unzipper.cpp

```diff
@@ -0,0 +1,217 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+
+// Interface header.
+#include "unzipper.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exception.h"
+#include "foundation/utility/minizip/unzip.h"
+
+// Boost headers.
+#include "boost/filesystem.hpp"
+
+// Standard headers.
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <sstream>
+
+using namespace std;
+namespace bf = boost::filesystem;
+
+namespace foundation
+{
+
+const int BUFFER_SIZE = 4096;
+
+const char* UnzipException::what() const throw() 
+{
+    stringstream exc_message;
+    exc_message << message;
+
+    if (err != 0)
+        exc_message << err;
+
+    return exc_message.str().c_str();
+}
+
+bool exists(const string& dirname) {
+    bf::path dir(dirname);
+    return bf::exists(dir);
+}
+
+bool make_dir(const string& dirname) 
+{
+    bf::path dir(dirname);
+    return bf::create_directories(dir);
+}
+
+int remove_dir(const string& dirname)
+{
+    bf::path dir(dirname);
+    return bf::remove_all(dir);
+}
+
+bool is_directory(const string& dirname) 
+{
+    // used own implementation of is_directory instead of boost implementation
+    // because this directory is not in filesystem, but in zipfile
+    return dirname[dirname.size() - 1] == '/';
+}
+
+void open_current_file(unzFile& zip_file) 
+{
+    const int err = unzOpenCurrentFile(zip_file);
+    if (err != UNZ_OK) 
+        throw UnzipException("Can't open file inside zip: ", err);
+}
+
+int read_chunk(unzFile& zip_file, char* buffer, const int chunk_size) {
+    const int err = unzReadCurrentFile(zip_file, buffer, chunk_size);
+
+    if (err == UNZ_ERRNO)
+        throw UnzipException("IO error while reading from zip");
+    if (err < 0)
+        throw UnzipException("zLib error with decompessing file: ", err);
+
+    return err;
+}
+
+void close_current_file(unzFile& zip_file) {
+    const int err = unzCloseCurrentFile(zip_file);
+
+    if (err == UNZ_CRCERROR)
+        throw UnzipException("CRC32 is not good");
+}
+
+string read_filename(unzFile& zip_file) 
+{
+    char filename_inzip[256];
```

Comment:
Seems like you should be able to know in advance the length of the filename:
http://stackoverflow.com/questions/16386335/how-does-one-read-files-in-subfolders-with-minizip